File: hard/straight_lines.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

type point struct {
	x int
	y int
}

func inLine(a, b, c point) bool {
	dx1, dy1, dx2, dy2 := a.x-b.x, a.y-b.y, a.x-c.x, a.y-c.y
	return dx1*dy2 == dx2*dy1
}

func line(a, b int, t []point) (l []int) {
	for i := 0; i < len(t); i++ {
		if i == a || i == b {
			l = append(l, i)
			continue
		}
		if inLine(t[i], t[a], t[b]) {
			if i < b {
				return l
			}
			l = append(l, i)
		}
	}
	return l
}

func main() {
	var x, y int
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " | ")
		var t []point
		for _, i := range s {
			fmt.Sscanf(i, "%d %d", &x, &y)
			t = append(t, point{x, y})
		}
		var r [][]int
		for i := 0; i < len(t)-2; i++ {
			for j := i + 1; j < len(t)-1; j++ {
				if l := line(i, j, t); len(l) > 2 {
					r = append(r, l)
				}
			}
		}
		fmt.Println(len(r))
	}
}

File: hard/straight_lines.rb
Point = Struct.new(:x, :y)

def in_line(a, b, c)
  return (a.x - b.x) * (a.y - c.y) == (a.x - c.x) * (a.y - b.y)
end

def line(ax, bx, t)
  t.each_with_index do |i, ix|
    return ix >= bx if ix != ax and ix != bx and in_line(i, t[ax], t[bx])
  end
  false
end

File.open(ARGV[0]).each_line do |line|
  s = line.split(" | ").map { |i|
    t = i.split().map(&:to_i)
    Point.new(t[0], t[1])
  }
  r = 0
  for ix in (0..s.length-2)
    for jx in (ix+1..s.length-1)
      r += 1 if line(ix, jx, s)
    end
  end
  puts r
end
